applehat/src/ui.rs

```diff
@@ -0,0 +1,285 @@
+        {
+            // Show currently selected value on alphanumeric display.
+            if let Some(value) = get_property(
+                &devices,
+                selected_device_id,
+                selected_node_id,
+                &self.selected_property_id,
+            ) {
+                print_str_decimal(
+                    &mut self.alphanum,
+                    &value,
+                    if self.selected_property_id == HUMIDITY_PROPERTY_ID {
+                        Some('%')
+                    } else {
+                        None
+                    },
+                );
+            } else {
+                self.alphanum.print_str("gone", false);
+            }
+        } else {
+            self.alphanum.print_str("    ", false);
+        }
+        if let Err(e) = self.alphanum.show() {
+            error!("Error displaying: {}", e);
+        }
+    }
+
+    fn button_pressed(&mut self, button_index: usize) {
+        debug!("Button {} pressed.", button_index);
+        match button_index {
+            0 => {
+                // Select next node.
+                let devices = self.controller.devices();
+                let nodes = find_nodes(&devices);
+                if !nodes.is_empty() {
+                    let new_index = if let Some(selected_node_id) = &self.selected_node_id {
+                        if let Some(current_index) = nodes
+                            .iter()
+                            .position(|(_, node_id, _)| node_id == selected_node_id)
+                        {
+                            (current_index + 1) % nodes.len()
+                        } else {
+                            0
+                        }
+                    } else {
+                        0
+                    };
+                    self.selected_node_id = Some(nodes[new_index].1.to_string());
+                }
+            }
+            1 => {
+                // Select next property.
+                let current_index = PROPERTY_IDS
+                    .iter()
+                    .position(|x| x == &self.selected_property_id)
+                    .unwrap_or(0);
+                self.selected_property_id =
+                    PROPERTY_IDS[(current_index + 1) % PROPERTY_IDS.len()].to_string();
+            }
+            2 => {
+                // Change brightness.
+                let current_index = BRIGHTNESS_LEVELS
+                    .iter()
+                    .position(|&level| level == self.selected_brightness)
+                    .unwrap_or(0);
+                self.selected_brightness =
+                    BRIGHTNESS_LEVELS[(current_index + 1) % BRIGHTNESS_LEVELS.len()];
+            }
+            _ => {}
+        }
+        self.update_display();
+    }
+
+    fn update_button_state(&mut self, new_state: [bool; 3]) {
+        for i in 0..3 {
+            if new_state[i] && !self.button_state[i] {
+                self.button_pressed(i);
```

Comment:
Are the buttons debounced in hardware or something?

Reply:
I guess so? It seems to work, anyway.